lib/ruby_lsp/requests/support/dependency_detector.rb

```diff
@@ -2,54 +2,64 @@
 # frozen_string_literal: true
 
 module RubyLsp
-  module DependencyDetector
-    class << self
-      extend T::Sig
-
-      sig { returns(String) }
-      def detected_formatter
-        # NOTE: Intentionally no $ at end, since we want to match rubocop-shopify, etc.
-        if direct_dependency?(/^rubocop/)
-          "rubocop"
-        elsif direct_dependency?(/^syntax_tree$/)
-          "syntax_tree"
-        else
-          "none"
-        end
-      end
+  class DependencyDetector
+    extend T::Sig
+
+    sig { void }
+    def initialize
+      @dependency_keys = T.let(nil, T.nilable(T::Array[String]))
```

Comment:
I think we can move a lot of things into this initializer and eagerly discover everything we need, so that we don't have to constantly grep on an array.

For example

```ruby
keys = dependency_keys

@detected_formatter = ... # check which formatter exists and set it
@has_typechecker = ...
@test_framework = ...
```

I also think this class might be safe enough to turn into a singleton. That way, we don't have to worry about passing it as arguments. E.g.:

```ruby
class DependencyDetector
  include Singleton
  # ...
end

# In requests
DependencyDetector.instance.has_typechecker?
```